modules/c++/mem/include/mem/ScratchMemory.hpp

```diff
@@ -52,7 +56,83 @@ inline void ScratchMemory::put<sys::ubyte>(const std::string& key,
     }
     mSegments.insert(
             iterSeg,
-            std::make_pair(key, Segment(numElements, numBuffers, alignment)));
+            std::make_pair(key, Segment(numElements, numBuffers, alignment, segmentOffset)));
+
+    mKeyOrder.push_back(key);
+}
+
+inline void ScratchMemory::release(const std::string& key)
+{
+    std::map<std::string, Segment>::const_iterator iterSeg = mSegments.find(key);
+    mReleasedKeys.insert(key);
+
+    if (mKeyOrder.back() == key)
+    {
+        const Segment& segment = iterSeg->second;
+        mOffset = segment.offset;
+    }
+    else
+    {
+        const Segment& segment = iterSeg->second;
+        mOffset = segment.offset;
+
+        std::vector<std::string>::iterator keyIter = std::find(mKeyOrder.begin(),
+                                                                     mKeyOrder.end(),
+                                                                     key);
+        std::vector<std::string>::iterator nextKeyIter = mKeyOrder.erase(keyIter);
+        mKeyOrder.push_back(key);
+
+        bool keepGoing = true;
+        std::string firstReleasedKey;
+        bool firstReleasedKeyFound = false;
```

Comment:
Can we remove `firstReleasedKeyFound` and use `firstReleasedKey.empty()` instead?